### PR TITLE
Refactor: No more than 1 namespace declaration per file

### PR DESCRIPTION
**Pull request by weswigham**
This refactors our codebase to conform to and adds a lint rule to enforce a file having at most one top level namespace declaration. This makes the translation into modules _much_ easier (and this refactor wasn't easily automatable, as some split namespaces existed to accomplish scope workarounds, which had to be done differently), as it means I don't have to track scope conflicts or what's exported from where when transforming the namespaces.

The primary effect of this is that a bunch of core utility files now have a clean split between their `@internal` and public halves. Hopefully this'll have the beneficial side-effect of making it much more obvious what's public vs what isn't when you're authoring new utilities.

This also slightly changes how we do some shim stuff in the harness, so that code will be more stable when it migrates from namespaces to modules - namely, no longer relying on `namespace Foo` making a global `Foo` to patch things into the global scope.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot pack this just to validate that renaming core.ts didn't break package generation or anything

File: src/testRunner/unittests/tscWatch/emitAndErrorUpdates.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 namespace ts.tscWatch {
     describe("unittests:: tsc-watch:: Emit times and Error updates in builder after program changes", () => {
         const config: File = {
-            path: `${projectRoot}/tsconfig.json`,
+            path: `${tscWatch.projectRoot}/tsconfig.json`,
```
Comment (sandersn) [inline src/testRunner/unittests/tscWatch/emitAndErrorUpdates.ts]: why is this needed if we are already in the ts.tscWatch namespace?

**Comment by weswigham** [comment]
@typescript-bot pack this now that lint's fixed

File: src/testRunner/unittests/tscWatch/emitAndErrorUpdates.ts (hunk 1 above)
Comment (weswigham) [reply src/testRunner/unittests/tscWatch/emitAndErrorUpdates.ts]: it's now gone~

File: src/compiler/builderPublic.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
+namespace ts {
```
Comment (sandersn) [inline src/compiler/builderPublic.ts]: x.ts/xPublic.ts is a fine naming scheme except that it makes bash-style completions worse because bu<TAB> doesn't produce a filename.

One solution is to use a public- prefix, which also seems bad. Or maybe a subdirectory. I don't really have *any* good ideas. Maybe this situation is temporary? I hope?

File: src/services/globalThisShim.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        window.globalThis = window;
+    }
+})());
+// #endregion The polyfill ends here.
+
+// if `process` is undefined, we're probably not running in node - patch legacy members onto the global scope
+// @ts-ignore
+if (typeof process === "undefined" || process.browser) {
+    /// TODO: this is used by VS, clean this up on both sides of the interface
+    //@ts-ignore
```
Comment (sandersn) [inline src/services/globalThisShim.ts]: mixed feelings about ts-ignore in ts file in Typescript codebase...

File: src/services/utilities.ts [hunk 4]
```diff
@@ -8,9 +8,10 @@ interface PromiseConstructor {
 /* @internal */
 declare let Promise: PromiseConstructor;
 
-// These utilities are common to multiple language service features.
 /* @internal */
 namespace ts {
+    // These utilities are common to multiple language service features.
+    //#region
```
Comment (sandersn) [inline src/services/utilities.ts]: does VS Code support `#region` ??

Comment (weswigham) [reply src/services/utilities.ts]: Yea

File: src/services/globalThisShim.ts (hunk 3 above)
Comment (weswigham) [reply src/services/globalThisShim.ts]: Eh, we're doing things that are untyped here (and the types "fix" is actually declaring these fields as globals, which I'd rather not do), so we need the bailout. It's also not our first - we already have one in the mapShim file to suppress an "unused type parameter" error (and is needed because said interface later merges with another in a dependent project where the param is actually used).

File: src/harness/harnessGlobals.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
+// Block scoped definitions work poorly for global variables, temporarily enable var
+/* eslint-disable no-var */
+
+// this will work in the browser via browserify
+var _chai: typeof chai = require("chai");
```
[the remark this answers is not in the document]
Comment (weswigham) [reply src/harness/harnessGlobals.ts]: But this is all test harness setup...?

File: src/compiler/builderPublic.ts (hunk 2 above)
Comment (sheetalkamat) [reply src/compiler/builderPublic.ts]: I like @sandersn 's idea of subfolder for internal. If we are not making subfolder i would like having `xInternal` and `x` so the file names exposes at smaller scale are nicer and without public

File: src/server/tsconfig.json [hunk 6]
```diff
@@ -15,7 +15,9 @@
     ],
     "files": [
         "types.ts",
+        "utilitiesPublic.ts",
         "utilities.ts",
+        "watchType.ts"
```
Comment (amcasey) [inline src/server/tsconfig.json]: @weswigham Comma
